service/history/shard/controller_test.go

```diff
@@ -626,6 +628,124 @@ func (s *controllerSuite) TestShardExplicitUnloadCancelGetOrCreate() {
 	s.Less(time.Since(start), 500*time.Millisecond)
 }
 
+func (s *controllerSuite) TestShardControllerFuzz() {
+	s.config.NumberOfShards = 10
+
+	s.mockServiceResolver.EXPECT().AddListener(shardControllerMembershipUpdateListenerName, gomock.Any()).Return(nil).AnyTimes()
+	s.mockServiceResolver.EXPECT().RemoveListener(shardControllerMembershipUpdateListenerName).Return(nil).AnyTimes()
+	s.mockClusterMetadata.EXPECT().GetCurrentClusterName().Return(cluster.TestCurrentClusterName).AnyTimes()
+	s.mockClusterMetadata.EXPECT().GetAllClusterInfo().Return(cluster.TestSingleDCClusterInfo).AnyTimes()
+
+	var engineStarts, engineStops int64
+	var getShards, closeContexts int64
+	var countCloseWg sync.WaitGroup
+
+	for shardID := int32(1); shardID <= s.config.NumberOfShards; shardID++ {
+		queueAckLevels := s.queueAckLevels()
+		queueStates := s.queueStates()
+
+		s.mockServiceResolver.EXPECT().Lookup(convert.Int32ToString(shardID)).Return(s.hostInfo, nil).AnyTimes()
+		s.mockEngineFactory.EXPECT().CreateEngine(contextMatcher(shardID)).DoAndReturn(func(shard Context) Engine {
+			mockEngine := NewMockEngine(s.controller)
+			status := new(int32)
+			mockEngine.EXPECT().Start().Do(func() {
+				if !atomic.CompareAndSwapInt32(status, common.DaemonStatusInitialized, common.DaemonStatusStarted) {
+					return
+				}
+				atomic.AddInt64(&engineStarts, 1)
+			}).AnyTimes()
+			mockEngine.EXPECT().Stop().Do(func() {
+				if !atomic.CompareAndSwapInt32(status, common.DaemonStatusStarted, common.DaemonStatusStopped) {
+					return
+				}
+				atomic.AddInt64(&engineStops, 1)
+			}).AnyTimes()
+			return mockEngine
+		}).AnyTimes()
+		s.mockShardManager.EXPECT().GetOrCreateShard(gomock.Any(), getOrCreateShardRequestMatcher(shardID)).DoAndReturn(
+			func(ctx context.Context, req *persistence.GetOrCreateShardRequest) (*persistence.GetOrCreateShardResponse, error) {
+				atomic.AddInt64(&getShards, 1)
+				countCloseWg.Add(1)
+				go func(ctx context.Context) {
+					<-ctx.Done()
+					atomic.AddInt64(&closeContexts, 1)
+					countCloseWg.Done()
+				}(req.LifecycleContext)
+				return &persistence.GetOrCreateShardResponse{
+					ShardInfo: &persistencespb.ShardInfo{
+						ShardId:                shardID,
+						Owner:                  s.hostInfo.Identity(),
+						RangeId:                5,
+						ReplicationDlqAckLevel: map[string]int64{},
+						QueueAckLevels:         queueAckLevels,
+						QueueStates:            queueStates,
+					},
+				}, nil
+			}).AnyTimes()
+		s.mockShardManager.EXPECT().UpdateShard(gomock.Any(), gomock.Any()).Return(nil).AnyTimes()
+	}
+
+	randomLoadedShard := func() (int32, Context) {
+		s.shardController.Lock()
+		defer s.shardController.Unlock()
+		if len(s.shardController.historyShards) == 0 {
+			return -1, nil
+		}
+		n := rand.Intn(len(s.shardController.historyShards))
+		for id, shard := range s.shardController.historyShards {
+			if n == 0 {
+				return id, shard
+			}
+			n--
+		}
```

Comment:
range over map is random anyway, we could just return the first one from map iterator.

Reply:
Not random enough!
https://stackoverflow.com/a/41020528
https://medium.com/i0exception/map-iteration-in-go-275abb76f721
https://dev.to/wallyqs/gos-map-iteration-order-is-not-that-random-mag